### DISCUSSION TITLE
Export subdiv settings per mesh

### DISCUSSION THREAD

**Discussion by flord** [Autodesk/maya-usd#2083] [Q&A] [answered]
Hi,
Is there a way to control the subdivision settings per object/prim? I see there is a setting called Subdivision Method in the export options, but it seems to affect all the objects that are exported. I need a way to control this per object.
There are a lot of settings related to subdivision on the shapes in Maya. I must confess I didn't try them all. But I played with a few and none seems to have an effet on the export. 

Thanks!

F

**Comment by womanyoyoyo-adsk** [comment, marked as answer]
Hi F,

Currently there is no way to control the subdivision setting per object when exporting to USD. You have many objects that need different USD subdivision methods when you're exporting?

Natalia

**Comment by flord** [comment]
Yes, we have assets containing objects that need subdivision and others 
that should remain unsubdivided. When we render with Arnold, we can 
control this per mesh.

Thanks!

**Reply by sharktacos** [reply]
If you set the Subdivisions in the Arnold section of the Mesh, this will be exported to USD (using the Maya USD export) and will appear in the Extra Attributes as “primvars arnold subdiv type.” In this way, subdivs can be set per object.

When rendering with Arnold, it appears that the regular USD subDiv settings have no effect. It would be good to get the Arnold and Maya plugins to "talk to each other" in this regard.

**Comment by kxl-adsk** [comment]
@flord There is something that may help you. If you dig into our commands documentation you can find
> defaultMeshScheme ... Sets the default subdivision scheme for exported Maya meshes, if the `USD_subdivisionScheme` attribute is not present on the Mesh. ...

Now, there are many ways how you could set this attribute...one USD-friendly way is via adaptors. You can read more in [here](https://github.com/Autodesk/maya-usd/blob/211287f24f4b3858257ea408c243f3ec3f76f14e/plugin/pxr/doc/README.md#registered-metadata-and-schema-attributes), but in short, you could do:
```
from pxr import UsdGeom, UsdMaya
UsdMaya.Adaptor("|my|mesh|shape").GetSchema(UsdGeom.Mesh).CreateAttribute(UsdGeom.Tokens.subdivisionScheme).Set(UsdGeom.Tokens.loop) # set subdivisionScheme=loop
```

PS - I have not tested it, so please don't hesitate to come back with questions.

**Comment by fabal** [comment]
Hi there, we (Animal Logic) have similar scenarios and this feature is a blocker for our transition to ADSK's plugin. `Adaptors` look promising though, we'll investigate shortly.

**Comment by senlegio** [comment]
Hey, i know it's been a long time but i came up with this python script with the help of ChatGPT.

```

import maya.cmds as cmds

def add_subdivision_attribute(shape_node):
    # Check if the attribute already exists
    if cmds.attributeQuery('USD_subdivisionScheme', node=shape_node, exists=True):
        # Delete the existing attribute
        cmds.deleteAttr(shape_node + ".USD_subdivisionScheme")

    # Check if Smooth Mesh Preview is on
    if cmds.attributeQuery('displaySmoothMesh', node=shape_node, exists=True):
        smooth_preview_on = cmds.getAttr(shape_node + ".displaySmoothMesh")
    else:
        # If the attribute doesn't exist, assume it's off
        smooth_preview_on = False

    # Add the custom attribute with the specified value
    cmds.addAttr(shape_node, ln="USD_subdivisionScheme", dt="string")
    cmds.setAttr(shape_node + ".USD_subdivisionScheme", "catmullClark" if smooth_preview_on else "none", type="string")

def process_hierarchy(node):
    # Recursively process children
    children = cmds.listRelatives(node, children=True, fullPath=True) or []
    for child in children:
        process_hierarchy(child)

    # Process the current node
    # Get the shape nodes of the object
    shapes = cmds.listRelatives(node, shapes=True, fullPath=True) or []
    for shape_node in shapes:
        add_subdivision_attribute(shape_node)

def scan_scene():
    # Get all objects in the scene
    all_objects = cmds.ls(dag=True, long=True)

    # Process each object and its hierarchy
    for obj in all_objects:
        process_hierarchy(obj)

def main():
    # Scan the entire scene
    scan_scene()

if _name_ == "_main_":
    main()

```

just run this script before you export. 

what this script does is:
1. scan all object and check their "displaySmoothMesh" attribute. (when you press 1 or 3 for your mesh, this attribute gets modified)
2. adding "USD_subdivisionScheme" extra attribute to all of the object.
3. set the proper value for "USD_subdivisionScheme" attribute based on their value of "displaySmoothMesh"

on export, you don't need to care on changing the subdivision mode anymore.
